tokio/src/sync/mpsc/bounded.rs

```diff
@@ -230,6 +230,74 @@ impl<T> Receiver<T> {
         poll_fn(|cx| self.chan.recv(cx)).await
     }
 
+    /// Receives the next values for this receiver and extends `buffer`.
+    ///
+    /// This method returns the number of values added to `buffer`.
+    ///
+    /// If there are no messages in the channel's queue, but the channel has
+    /// not yet been closed, this method will sleep until a message is sent or
+    /// the channel is closed.  Note that if [`close`] is called, but there are
+    /// still outstanding [`Permits`] from before it was closed, the channel is
+    /// not considered closed by `recv` until the permits are released.
+    ///
+    /// This method will never return 0 unless the channel has been closed and
+    /// there are no remaining messages in the channel's queue. This indicates
+    /// that no further values can ever be received from this `Receiver`. The
+    /// channel is closed when all senders have been dropped, or when [`close`]
+    /// is called.
+    ///
+    /// If at the time of the call `buffer` has unused capacity,
+    /// `recv_many` extends the buffer with no more elements than
+    /// its unused capacity. Should `buffer` have no initial unused
+    /// capacity, additional elements are first reserved.
```

Comment:
I was looking at this paragraph, and I think it could be improved like this:
```suggestion
    /// If `buffer` has unused capacity, then this call will not reserve
    /// additional space in `buffer`. This means that the maximum number of
    /// received messages is `buffer.capacity() - buffer.len()`. However, if
    /// the capacity is equal to the length, then this call will increase the
    /// capacity to make space for additional elements.
```
This actually raises a question: Perhaps it makes sense to only reserve space when we return a message? This way, we don't consume extra memory until a message arrives, and if the channel gets closed, we don't reserve any space.

What do you think?

Reply:
I like your language here better.

I also like the idea of not allocating memory unless needed.  If the channels are very lightweight perhaps there are often created and a finite number of messages sent and then closed.  Why force the caller to always allocate one more message than is actually sent?

In chan.rs, `recv_many`:
```
let mut insufficient_capacity = buffer.capacity() == buffer.len();
...
                () => {
                    while (buffer.len() < buffer.capacity() || insufficient_capacity) {
                        match rx_fields.list.pop(&self.inner.tx) {
                            Some(Read::Value(value)) => {
                                if insufficient_capacity {
                                    buffer.reserve(super::BLOCK_CAP);
                                    insufficient_capacity = false;
                                }
                                buffer.push(value);
                            }
```

Reply:
Does it make sense to urge users to manage the buffer's capacity outside `recv_many`?
e.g.,

If `buffer` has unused capacity, then this call will not reserve
additional space in `buffer`. This means that the maximum number of
received messages is `buffer.capacity() - buffer.len()`.  
Efficient client code should ensure `buffer` has unused capacity,
but if the capacity is equal to the length and there is at least one message
in the channel's queue, then this call will _conservatively_ increase the capacity
to make space for additional elements.